src/librbd/ObjectMap.h

```diff
@@ -23,7 +24,7 @@ struct BlockGuardCell;
 class ImageCtx;
 
 template <typename ImageCtxT = ImageCtx>
-class ObjectMap {
+class ObjectMap : public RefCountedObject {
```

Comment:
Also, it would be better if the change is not difficult to inherit from `RefCountedObjectSafe` (here and elsewhere) so that you can protect against manual ref count changes like this. It's better to exclusively use smart pointers to modify the reference count. See also the comment above `RefCountedObject` for explanation.

Reply:
As mentioned below, this would require overhaul of a bunch of code since currently objects are created with `new` keyword. I could do a follow-up PR to change that pattern after fixing race conditions in the code, if that sounds okay.

Reply:
Up to you and @dillaman. It's just a suggestion for defensive programming.

Reply:
> currently objects are created with `new` keyword

I'll admit it's been a while since I deeply reviewed this PR (and obviously a lot has changed w/ the reference counting helpers in the meantime), but why would the `new` keyword prevent Patrick's suggestion?

Reply:
`RefCountedObjectSafe` prevents direct access to `get()` and `put()` methods. If objects are created using `ceph::ref_t` (boost::instrusive_ptr), then it inherently has reference counting mechanism and doesn't need direct access to those methods. Maybe I worded it wrongly, it's not necessarily the `new` keyword (which is still used with intrusive_ptr), but the lack of `ceph::ref_t`. So for objects without that, ref counting can only happen with access to `get()` and `put()` methods. Therefore, at the moment current code only inherits from `RefCountedObject`

I could however use the below suggestion and apply it where it's possible.